cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go

```diff
@@ -39,67 +39,44 @@ const (
 	NodesKubeadmConfigClusterRoleName = "kubeadm:nodes-kubeadm-config"
 )
 
+type clusterStatusMutator func(*kubeadmapi.ClusterStatus) error
+
 // ResetClusterStatusForNode removes the APIEndpoint of a given control-plane node
 // from the ClusterStatus and updates the kubeadm ConfigMap
 func ResetClusterStatusForNode(nodeName string, client clientset.Interface) error {
 	fmt.Printf("[reset] Removing info for node %q from the ConfigMap %q in the %q Namespace\n",
 		nodeName, kubeadmconstants.KubeadmConfigConfigMap, metav1.NamespaceSystem)
 
-	// Get the kubeadm ConfigMap
-	configMap, err := client.CoreV1().ConfigMaps(metav1.NamespaceSystem).Get(kubeadmconstants.KubeadmConfigConfigMap, metav1.GetOptions{})
-	if err != nil {
-		return errors.Wrap(err, "failed to get config map")
-	}
-
-	// Handle missing ClusterConfiguration in the ConfigMap. Should only happen if someone manually
-	// interacted with the ConfigMap.
-	clusterConfigurationYaml, ok := configMap.Data[kubeadmconstants.ClusterConfigurationConfigMapKey]
-	if !ok {
-		return errors.Errorf("cannot find key %q in ConfigMap %q in the %q Namespace",
-			kubeadmconstants.ClusterConfigurationConfigMapKey, kubeadmconstants.KubeadmConfigConfigMap, metav1.NamespaceSystem)
-	}
+	return apiclient.MutateConfigMap(client, metav1.ObjectMeta{
```

Comment:
This seems overly complicated form of what is typically referred to as CreateOrUpdate, I don't think it needs to be a closure at this level

Reply:
The problem with `CreateOrUpdate` is that it involves some boilerplate that needs repeating both in reset and when joining new control plane nodes (unmarshal cluster status, mutate, marshal cluster status back). Also, the whole repeat if conflict would need to be repeated in both places.

`CreateOrMutateConfigMap` and `MutateConfigMap` approach this in a different way, you just implement the mutator and `MutateConfigMap` will take care of:

1. Fetch the resource
2. Callback to the mutator
3. Update
  3.1. If conflict, go to 1

Reply:
However, I will look into `RetryOnConflict` as @detiber suggested, it's exactly the same functionality that I wrote here.

Reply:
k, I see why you're doing it this way now.  Thanks for the explanation.